File: techniques/sign-test/r/sign_test.R
# Sign Test (Reference §6.1)
# Base R has stats::binom.test (the engine for the sign test).
# Run with:  Rscript sign_test.R
#
# Inputs:
#   x, m0       : sample and null median
#   x1, x2      : paired samples (test the median of the differences)
#   alternative : "two.sided" / "greater" / "less"

sign_test_scratch <- function(x, m0 = 0, alternative = "two.sided") {
  n_pos <- sum(x > m0); n_neg <- sum(x < m0); n_tie <- sum(x == m0)
  n_eff <- n_pos + n_neg
  if (n_eff == 0) return(list(S_pos = 0, n_effective = 0, p_value = 1))
  p_value <- binom.test(n_pos, n_eff, p = 0.5, alternative = alternative)$p.value
  list(S_pos = n_pos, n_neg = n_neg, n_tie = n_tie,
       n_effective = n_eff, p_value = p_value, method = "sign test")
}

paired_sign_test_scratch <- function(x1, x2, alternative = "two.sided") {
  stopifnot(length(x1) == length(x2))
  sign_test_scratch(x1 - x2, m0 = 0, alternative = alternative)
}

# Library: BSDA::SIGN.test (one- and paired-sample sign test)
library_demo <- function(x, m0) {
  if (requireNamespace("BSDA", quietly = TRUE))
    print(BSDA::SIGN.test(x, md = m0, alternative = "greater"))
  else cat("(install 'BSDA' for SIGN.test; the from-scratch matches binom.test)\n")
}

if (sys.nframe() == 0) {
  set.seed(0)
  cat("=== One-sample, median > 50 ===\n")
  x <- rnorm(25, 52, 10); print(sign_test_scratch(x, m0 = 50, alternative = "greater"))
  cat("\n=== Paired ===\n")
  pre <- rnorm(20, 100, 10); post <- pre + rnorm(20, 2, 4)
  print(paired_sign_test_scratch(post, pre, alternative = "greater"))
  cat("\n--- library ---\n"); library_demo(x, 50)
}

File: techniques/sign-test/python/sign_test.py
"""The Sign Test (Reference §6.1).

The minimal-assumption test about a median (or paired differences):

  One-sample : H0: median(x) = m0
               Let S = #{x_i > m0}; ignore ties.
               Under H0, S ~ Binomial(n_no_tie, 0.5).

  Paired     : Test that the median of (x1 - x2) is 0 -- a one-sample sign test
               on the differences.

Only assumes continuous, independent observations. Compared to Wilcoxon
signed-rank (techniques/wilcoxon-signed-rank), the sign test uses LESS
information about the differences (only their signs, not their magnitudes),
so it has lower power but also fewer assumptions (no symmetry needed).
"""
from __future__ import annotations

import math
from typing import Sequence

from scipy import stats


def sign_test(x: Sequence[float], m0: float = 0.0,
              alternative: str = "two-sided") -> dict:
    """One-sample sign test of H0: median(x) = m0.

    ``alternative``: "two-sided" / "greater" / "less".
    """
    n_pos = sum(1 for v in x if v > m0)
    n_neg = sum(1 for v in x if v < m0)
    n_tie = sum(1 for v in x if v == m0)
    n_eff = n_pos + n_neg
    if n_eff == 0:
        return {"S": 0, "n_effective": 0, "p_value": 1.0, "method": "sign test"}

    if alternative == "two-sided":
        # scipy's binomtest two-sided 'method of small p-values'
        p_value = float(stats.binomtest(n_pos, n_eff, 0.5).pvalue)
    elif alternative == "greater":
        p_value = float(stats.binomtest(n_pos, n_eff, 0.5, alternative="greater").pvalue)
    elif alternative == "less":
        p_value = float(stats.binomtest(n_pos, n_eff, 0.5, alternative="less").pvalue)
    else:
        raise ValueError("alternative must be 'two-sided', 'greater', or 'less'")

    return {"S_pos": n_pos, "n_neg": n_neg, "n_tie": n_tie,
            "n_effective": n_eff, "p_value": p_value, "method": "sign test"}


def paired_sign_test(x1: Sequence[float], x2: Sequence[float],
                     alternative: str = "two-sided") -> dict:
    """Paired sign test: H0: median(x1 - x2) = 0."""
    if len(x1) != len(x2):
        raise ValueError("paired test requires equal-length samples")
    d = [a - b for a, b in zip(x1, x2)]
    return sign_test(d, m0=0.0, alternative=alternative)


def library_versions(x, m0=0.0):
    try:
        from statsmodels.stats.descriptivestats import sign_test as sm_sign
        stat, p = sm_sign(x, mu0=m0)
        return {"statsmodels sign_test": (float(stat), float(p))}
    except Exception as exc:
        return {"statsmodels": f"error: {exc}"}


if __name__ == "__main__":
    import numpy as np
    rng = np.random.default_rng(0)

    print("=== One-sample (median > 50?) ===")
    x = rng.normal(52, 10, 25).tolist()
    print(f"  observed median = {sorted(x)[len(x)//2]:.2f}")
    for k, v in sign_test(x, m0=50, alternative="greater").items():
        print(f"  {k:14s}: {v}")

    print("\n=== Paired (post > pre?) ===")
    pre = rng.normal(100, 10, 20).tolist()
    post = [p + rng.normal(2, 4) for p in pre]
    for k, v in paired_sign_test(post, pre, alternative="greater").items():
        print(f"  {k:14s}: {v}")

    print("\n--- library (statsmodels sign_test) ---")
    for k, v in library_versions(x, m0=50).items():
        print(f"  {k}: {v}")
